File: Macro-Processor-main/algo1/pass1_m.py
# input file
input_file = 'input2.txt'

# open and read input file
fp = open(input_file, 'r')
data = fp.readline()

line_no = 1  # counter for line in input file
mntc = 1
mdtc = 1

# open/create other files
fp_mn = open('mnt_table.txt', 'w')
fp_md = open('mdt_table.txt', 'w')
fp_out = open('output_file1.txt', 'w')

while data != '':  # Check for end of file
    elements = data.split()
    if elements[1] == 'MACRO':
        data = fp.readline()
        line_no += 1

        # store macro arguments (for use in numbering)
        macro_arguments = data[data.find('&'):].rstrip().split(',')

        fp_mn.write('%d\t%s\t%d\n' %(mntc, data[data.find('-'):].rstrip(), mdtc))
        fp_md.write('%d\t%s\t\n' %(mdtc, data[data.find('-'):].rstrip()))
        mntc += 1; mdtc += 1
#        counter_mn = 0

        # read next line from input before executing while loop
        data = fp.readline()
        line_no += 1
        while elements[1] != 'MEND':  # while it is inside MACRO block
            # prepare string to be written
            start = data.find('-')
            end = data.find(',')
            string_to_write = '%d\t%s' %(mdtc, data[start:end+1].rstrip())

            if '&' in data[end:]:
                string_to_write += '#%d\n' %(macro_arguments.index(data[end+1:].rstrip()))
#                counter_mn += 1
            else:
                string_to_write += data[end+1:].split('&')[-1]

            fp_md.write(string_to_write)
            mdtc += 1

            # read next line in input
            data = fp.readline()
            line_no += 1
            elements = data.split()

        if 'MEND' in data:  # Macro END detected, write this line and read next line
            fp_md.write('%d\t%s' %(mdtc, data))
            mdtc += 1
            # read next line in input
            data = fp.readline()
            line_no += 1
    else:  # write to output file
        fp_out.write(data)
        data = fp.readline()
        line_no += 1

# close files
fp.close()
fp_mn.close()
fp_md.close()
fp_out.close()

File: Macro-Processor-main/algo1/pass1_m.c
#include<stdio.h>
#include<stdlib.h>
#include<string.h>

void main()
{
	FILE *f1, *f2, *f3, *f4;
	char label[100], code[100], operand[100];
	int mntc = 1, mdtc = 1, ala[100], i = 0, j = 0, k = 0, m = 0, counter = 0, flag = 0;
	
	f1 = fopen ("input.txt","r");
	f2 = fopen ("mnt_table.txt","w");
	f3 = fopen ("mdt_table.txt","w");
	f4 = fopen ("output_file1.txt","w");
	
	while (fscanf (f1,"%s%s%s",label,code,operand) != EOF)
	{
		if ( strcmp (code, "MACRO") == 0 )
		{	
			fscanf (f1,"%s%s%s",label,code,operand);
			fprintf (f2,"%d\t%s\t%s\t%s\t%d\n",mntc,label,code,operand,mdtc);
			fprintf (f3,"%d\t%s\t%s\t%s\n",mdtc,label,code,operand);
			mntc ++;
			mdtc ++;
			i = 0;
			j = 0;
			counter = 0;
			while ( operand[i] != '\0' )
			{
				if ( operand[i ++] == '&' )
					while (operand[i] != ',' && operand[i] != '\0')
						ala[j ++] = operand[i ++];
					ala[j ++] = ' ';
				i ++;
			}
			while ( !( strcmp (code, "MEND") == 0) )
			{
				i = 0;
				fscanf (f1,"%s%s%s",label,code,operand);
				while ( operand[i] != '\0' ) // substitue index notation from ala
				{
					if ( operand[i] == '&' )
					{
						counter = 0;
						m = i;
						flag = 0;
						++ i;
						for (k = 0; k < j; k ++)
						{
							if (ala[k] == operand[i])
							{
								flag = 1;
								i ++;
							}
							if (flag == 0)
								if (ala[k] == ' ')
									counter ++;
						}
							if (flag == 1)
							{
								operand[m] = '#';
								operand[++ m] = ( (char) counter ) + 48;
								operand[++ m] ='\0';
								i = m;
							}
						}
						else
							i ++;
				}
				fprintf (f3,"%d\t%s\t%s\t%s\n",mdtc,label,code,operand);
				mdtc ++;
			}
				
		}
		else			
			fprintf (f4,"%s\t%s\t%s\n",label,code,operand);
	}
	
	fclose (f1);
	fclose (f2);
	fclose (f3);
	fclose (f4);
}
